### PR TITLE
MWPW-148050: auto merge stage to main

### PR DESCRIPTION
**Pull request by mokimo**
### Description
Merge PRs from stage->main automatically given 4 sign-offs, as long as it's monday-thursday 8AM UTC to 8PM UTC

Resolves: [MWPW-148050](https://jira.corp.adobe.com/browse/MWPW-148050)

**Test URLs:**
- Before: https://main--milo--mokimo.hlx.page/?martech=off
- After: https://merge-process--milo--mokimo.hlx.live/?martech=off

### CODE REVIEW

**Review by overmyheadandbody — COMMENTED** [review]
Are we ready to merge this in? I was under the impression that we give the feature branch to stage merge a few weeks to figure out and address any issues and only then would we feel confident to do automatic merges to main. What timeline do you have in mind for this?

**Comment by mokimo** [comment]
@overmyheadandbody No specific timeline, I figured I can open the PR at any point, so this doesn't get lost and gives people the chance to give input.

Since this is the 2nd week with the stage automation, unless we see any big issues this should be good to be ready next week. This would also nicely fall in place with the milo deep dives on the 29th of May, where we can present all the changes to the process as a whole.

File: .github/workflows/merge-to-main.js
Comment (robert-bogos) [inline .github/workflows/merge-to-main.js:28]: I guess we are caching `isSunday`, `isSaturday` and `isFriday` just for readability. Otherwise we could do 
```suggestion
  return hour >= 8 && hour <= 20 && day !== 5 && day !== 6 && day !== 0;
```

**Review by robert-bogos — APPROVED** [review]

File: .github/workflows/merge-to-main.js
Comment (mokimo) [reply .github/workflows/merge-to-main.js:28]: I agree with the suggestion & had it like that before, but `0, 5, 6` is not immediately obvious compared to its not friday/saturday/sunday - hence you are spot on, it's a readability thing!

File: .github/workflows/merge-to-main.js
Comment (mokimo) [inline .github/workflows/merge-to-main.js:55]: We need to trigger the stage->main PR workflow after a merge

File: .github/workflows/merge-to-main.js
Comment (mokimo) [reply .github/workflows/merge-to-main.js:1]: Sent out the communication, if you didn't see it, let me know - then we have a big gap in our communication strategy, I've cross posted it a lot.

File: .github/workflows/merge-to-main.js
Comment (mokimo) [reply .github/workflows/merge-to-main.js:42]: In case of CSOs that's precisely what we don't want. We can adjust this if it doesn't work in its current form IMO.

Comment (narcis-radu) [reply .github/workflows/merge-to-main.js:42]: @mokimo - can you elaborate on that? Generally we're looking for sign offs from SOTs, but in the case of CSOs we will mostly do manual merges to speed up the fix.

Comment (mokimo) [reply .github/workflows/merge-to-main.js:42]: > CSO
Revert stage batch. Go on the merged PRs and click the "revert" button. (Notify folks to re-open their PRs afterwards)
Create PR to merge into stage with the fix or production-PR-revert
Get 2 approvals on it (or an admin can insta merge after tests are green)
Try to get the fix tested, but in a CSO case it's fine to simply apply the 4 SOT labels and merge into main.

As per the opening post in https://github.com/orgs/adobecom/discussions/2206

Comment (mokimo) [reply .github/workflows/merge-to-main.js:42]: With the new process, after you get approvals on a revert PR - merged it to stage - and then just apply the 4 SOT labels yourself, that enables everybody to step in to get the fix merged.

File: .github/workflows/helpers.js
Comment (narcis-radu) [reply .github/workflows/helpers.js:48]: @mokimo - what do you think about this? it's an easy change.

Comment (mokimo) [reply .github/workflows/helpers.js:48]: Yup will change that!

File: .github/workflows/merge-to-main.js
Comment (mokimo) [inline .github/workflows/merge-to-main.js:50]: We will need to merge main into stage, after merging stage into main. Otherwise, our stage branch is always going to be outdated which might lead to issues trying to automatically merge a PR unless we force merge it. We don't want to force merge to avoid any potential issues.

Stage:
Screenshot 2024-05-17 at 13 44 40

Main:
Screenshot 2024-05-17 at 13 42 10

Main receives an extra commit `e6d5f7c` due to the merge-commit. We will need to include `e6d5f7c` in stage as well, right after merging into main to get the two histories aligned.

Now as we'd merge `e6d5f7c` into stage, that would also always create a new commit, but this means stage is ahead of main, which is the normal case - so we'd still end up with an unnecessary commit but it's the lesser evil.

File: .github/workflows/merge-to-main.js
Comment (mokimo) [reply .github/workflows/merge-to-main.js:55]: Done

File: .github/workflows/helpers.js
Comment (mokimo) [reply .github/workflows/helpers.js]: Done

File: .github/workflows/merge-to-main.js
Comment (mokimo) [reply .github/workflows/merge-to-main.js:50]: Testing if that's really needed or an outdated branch doesn't stop the integration from merging.
I'd like to not back-merge if we can avoid it.

**Comment by mokimo** [comment]
https://github.com/adobecom/milo/pull/2362
image

Permissions are working, SOT labels are accounted for, this should be good to go

File: .github/workflows/merge-to-main.js
Comment (mokimo) [reply .github/workflows/merge-to-main.js:50]: Not needed. We can live with an outdated branch.

File: .github/workflows/helpers.js
Comment (robert-bogos) [inline .github/workflows/helpers.js:60]: Nit: this can be written shorter
```suggestion
  return RCPDates.some(({ start, end }) => start <= now && now <= end)
    ? (console.log('Current date is within a RCP. Stopping execution.'), true) : false;
```
Feel free to not make this change if you don't want to. It's less code, but might be less readable as it uses the less common [JS comma operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Comma_operator)

**Review by robert-bogos — APPROVED** [review]
Nice work, looks good from my side 👍

File: .github/workflows/helpers.js
Comment (overmyheadandbody) [inline .github/workflows/helpers.js:74]: Is this still needed?

File: .github/workflows/mark-stale-prs.yaml
Comment (overmyheadandbody) [inline .github/workflows/mark-stale-prs.yaml:1]: Not important, but I see we now have [names](https://github.com/adobecom/milo/blob/stage/.github/workflows/merge-to-stage.yaml#L1) without any quotes, [names](https://github.com/adobecom/milo/blob/stage/.github/workflows/mark-stale-prs.yaml#L1) with double quotes and now this. We should stick to a rule for consistency

File: .github/workflows/merge-to-main.js
Comment (overmyheadandbody) [inline .github/workflows/merge-to-main.js:28]: This sounds like a good start for IST/Europe. Do you see this expanding to the US timezone as well at one point?

File: .github/workflows/merge-to-main.yaml
Comment (overmyheadandbody) [inline .github/workflows/merge-to-main.yaml:5]: Judging by this and the JS file, we're looking just for SOT approvals and nothing else, right? Meaning: no more 'Ready for Main' label required. Is there any use-case where we would need to look out for the 'do not merge' label? I can't think of some particular use-cases right now, but I just saw the label on a [PR](https://github.com/adobecom/milo/pull/2362) of yours which gave me the idea.

**Review by overmyheadandbody — APPROVED** [review]

File: .github/workflows/helpers.js
Comment (mokimo) [reply .github/workflows/helpers.js:74]: Yes

**Review by narcis-radu — APPROVED** [review]

File: .github/workflows/mark-stale-prs.yaml
Comment (mokimo) [reply .github/workflows/mark-stale-prs.yaml:1]: There is no easy was to decide on this, for longer text with special characters, using (single) quotes is the way to go.
This won't work with numbers though, as it would convert numbers to strings.
Then there are smaller lines such as
```
    runs-on: ubuntu-latest
```

...

For the sake of simplicity, I'll just remove the quotes from the name for now and keep it on the cron schedule

Comment (mokimo) [reply .github/workflows/mark-stale-prs.yaml:1]: If we don't like the inconsistency we should do _one_ PR and potentially look into configuring the linter to handle a standard config for us in the future

File: .github/workflows/merge-to-main.js
Comment (mokimo) [reply .github/workflows/merge-to-main.js:28]: 8PM UTC would be:
- 4PM east coast (New York)
- 2PM (Utah)
- 1PM (San Jose)

I tried to account for an overlap of all (US) time zones without stretching them out too aggressively to give us good coverage in terms of available engineers. 

If you have any better suggestions, I'd gladly implement them - otherwise I'd recommend we start with trying to merge "earlier" into the day

File: .github/workflows/merge-to-main.yaml
Comment (mokimo) [reply .github/workflows/merge-to-main.yaml:5]: Yes, just the SOT approvals & a PR needs to be mergeable.
- Conflicts = PR can't be merged
- Requesting changes = PR can't be merged
- only 3/5 SOT labels = PR will not be merged

I thought about the `do not merge` label, but the standard GitHub workflow is requesting changes on a PR, which should be good enough

Comment (mokimo) [reply .github/workflows/merge-to-main.yaml:5]: `Ready for Main` was used to quickly filter/indicate the Stage->Main PRs, but as we have #milo-changelog + a recurring title, that should be good enough to find the prod PRs IMO

**Pull request merged**
